tests/cluster/tests/24-links.tcl

```diff
@@ -59,32 +66,38 @@ test "Disconnect link when send buffer limit reached" {
     set orig_link_p1_to_p2 [get_link_to_peer $primary1_id $primary2_name]
     set orig_link_p1_from_p2 [get_link_from_peer $primary1_id $primary2_name]
 
-    # On primary1, set cluster link send buffer limit to 32MB
+    # On primary1, set cluster link send buffer limit to 256KB, which is large enough to not be
+    # overflowed by regular gossip messages but also small enough that it doesn't take too much
+    # memory to overflow it. If it is set too high, Redis may get OOM killed by kernel before this
+    # limit is overflowed in some RAM-limited test environments.
     set oldlimit [lindex [$primary1 CONFIG get cluster-link-sendbuf-limit] 1]
-    $primary1 CONFIG set cluster-link-sendbuf-limit [expr 32*1024*1024]
+    $primary1 CONFIG set cluster-link-sendbuf-limit [expr 256*1024]
     assert {[get_info_field [$primary1 cluster info] total_cluster_links_buffer_limit_exceeded] eq 0}
 
     # To manufacture an ever-growing send buffer from primary1 to primary2,
     # make primary2 unresponsive.
     set primary2_pid [get_instance_attrib redis $primary2_id pid]
     exec kill -SIGSTOP $primary2_pid
 
-    # On primary1, send a 10MB Pubsub message. It will stay in send buffer of
-    # the link from primary1 to primary2
-    $primary1 publish channel [prepare_value [expr 10*1024*1024]]
-
-    # Check the same link has not been disconnected, but its send buffer has grown
-    set same_link_p1_to_p2 [get_link_to_peer $primary1_id $primary2_name]
-    assert {[dict get $same_link_p1_to_p2 create_time] eq [dict get $orig_link_p1_to_p2 create_time]}
-    assert {[dict get $same_link_p1_to_p2 send_buffer_allocated] > [dict get $orig_link_p1_to_p2 send_buffer_allocated]}
-
-    # On primary1, send another 30MB Pubsub message.
-    $primary1 publish channel [prepare_value [expr 30*1024*1024]]
+    # On primary1, send 128KB Pubsub messages in a loop until the send buffer of the link from
+    # primary1 to primary2 exceeds buffer limit therefore be dropped.
+    # For the send buffer to grow, we need to first exhaust TCP send buffer of primary1 and TCP
+    # receive buffer of primary2 first. The sizes of these two buffers vary by OS, but 100 128KB
+    # messages should be sufficient.
+    set i 0
+    wait_for_condition 100 0 {
+        [catch {incr i} e] == 0 &&
+        [catch {$primary1 publish channel [prepare_value [expr 128*1024]]} e] == 0 &&
+        [catch {after 500} e] == 0 &&
```

Comment:
Here is what I tried:

```
    while {1} {
        $primary1 multi
        $primary1 publish channel [prepare_value [expr 128*1024]]
        $primary1 cluster links
        set ret [$primary1 exec]
        set links [lindex $ret 1]
        set link_p1_to_p2 [get_link_to_peer_from_cluster_links_output $links $primary2_name]
        if {[dict get $link_p1_to_p2 send_buffer_allocated] > 256*1024} {
            break;
        }
    }

    wait_for_condition 10 500 {
        [get_info_field [$primary1 cluster info] total_cluster_links_buffer_limit_exceeded] eq 1
    } else {
        fail "Cluster link not freed as expected"
    }
```

This didn't work, because `send_buffer_allocated` being over the `cluster-link-sendbuf-limit` is not a reliable indicator that it will for sure be dropped. This is because when a 128KB message comes in, `jemalloc` over-allocate buffer and takes `send_buffer_allocated` over 256KB. But `send_buffer_allocated` will get shrunk below 256KB within `clusterNodeCronResizeBuffers` cron job, before being checked against `cluster-link-sendbuf-limit` in `clusterNodeCronFreeLinkOnBufferLimitReached`. (This is the correct ordering)

I tried replacing `send_buffer_allocated` with `send_buffer_used` but that makes the test flaky in the other direction - a link may get dropped without `send_buffer_allocated` being over 256KB, so once the we `break` out of the `while` loop, `total_cluster_links_buffer_limit_exceeded` is already more than 1.

Any other idea?

Reply:
I guess not, we can leave this as is.